torch/distributions/transforms.py

```diff
@@ -0,0 +1,422 @@
+import torch
+from torch.autograd import Variable
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all, lazy_property
+from torch.nn.functional import sigmoid
+
+__all__ = [
+    'AbsTransform',
+    'AffineTransform',
+    'BoltzmannTransform',
+    'ComposeTransform',
+    'ExpTransform',
+    'LowerCholeskyTransform',
+    'SigmoidTransform',
+    'StickBreakingTransform',
+    'Transform',
+    'identity_transform',
+]
+
+
+class Transform(object):
+    """
+    Abstract class for invertable transformations with computable log
+    det jacobians. They are primarily used in
+    :class:`torch.distributions.TransformedDistribution`.
+
+    Caching is useful for tranforms whose inverses are either expensive or
+    numerically unstable. Note that care must be taken with memoized values
+    since the autograd graph may be reversed. For example while the following
+    works with or without caching::
+
+        y = t(x)
+        t.log_abs_det_jacobian(x, y).backward()  # x will receive gradients.
+
+    However the following will error when caching due to dependency reversal::
+
+        y = t(x)
+        z = t.inv(y)
+        grad(z.sum(), [y])  # error because z is x
+
+    Derived classes should implement one or both of :meth:`_call` or
+    :meth:`_inverse`. Derived classes that set `bijective=True` should also
+    implement :meth:`log_abs_det_jacobian`.
+
+    Args:
+        cache_size (int): Size of cache. If zero, no caching is done. If one,
+            the latest single value is cached. Only 0 and 1 are supported.
+
+    Attributes:
+        domain (:class:`~torch.distributions.constraints.Constraint`):
+            The constraint representing valid inputs to this transform.
+        codomain (:class:`~torch.distributions.constraints.Constraint`):
+            The constraint representing valid outputs to this transform
+            which are inputs to the inverse transform.
+        bijective (bool): Whether this transform is bijective. A transform
+            ``t`` is bijective iff ``t.inv(t(x)) == x`` and
+            ``t(t.inv(y)) == y`` for every ``x`` in the domain and ``y`` in
+            the codomain. Transforms that are not bijective should at least
+            maintain the weaker pseudoinverse properties
+            ``t(t.inv(t(x)) == t(x)`` and ``t.inv(t(t.inv(y))) == t.inv(y)``.
+    """
+    bijective = False
+
+    def __init__(self, cache_size=0):
+        if cache_size == 0:
+            pass  # default behavior
+        elif cache_size == 1:
+            self._cached_x_y = None, None
+            self.__call__ = self._cached_call
+            self._inv_call = self._cached_inverse
+        else:
+            raise NotImplementedError('cache_size must be 0 or 1')
+
+    @lazy_property
```

Comment:
I don't think we want to make it a lazy property, or we'll end up with reference cycles. Creation of a new inverse should be very cheap, so let's just create a new one every time

Reply:
Are you sure? That will break the nice property that `t.inv.inv is t`?

Reply:
Done. At least we have `t.inv.inv == t`.

Reply:
Why would it break that property? We only ensure that inverse transforms have reference to the base one (and return it when asked for their inverse), while the base transform never holds any inverses. So `t.inv.inv is t`, it's just no longer true that `t.inv is t.inv` (not the same object by identity, but I'd say that comparing transforms by identity isn't a very good idea either).

Reply:
BTW we could keep `t.inv.inv is t` if only we kept a weakref to the inverse. Should be quite simple:
```python
@property
def inv(self):
  maybe_inv = self._inv()
  if maybe_inv is not None:
    return maybe_inv
  inv = InverseTransform(self)
  self._inv = weakref.ref(inv)
  return inv
```

Reply:
The property no longer holds for `InverseTransform`s:
```py
t = ExpTransform()
assert t.inv.inv is t      # <---- this works fine, as you observe
s = t.inv
assert s.inv.inv is not s  # <---- but this makes a second .inv copy
assert s.inv.inv == s      # <---- this is enough to make me happy :-)
```

Reply:
> weak ref

Hmm so `InverseTransform`s would have strong references to their "forward" transform, but forward transforms would have only weak references to inverse transforms? That sounds reasonable. Let me see if I can get that to work using your snippet.

Reply:
Thanks, @apaszke that ended up working well :smile: